### PR TITLE
stmhal: Fixed ESPRUINO_PICO build

### PR DESCRIPTION
**Pull request by dhylands**
The ESPRUINO_PICO only has 384K of flash, so this fixes it to use a different linker script.

With the current build, it was using a linker script for the 512K part and overflowing the .text section by 2.5K without failing to build (but producing an invalid image).

I also rearraged the contents of the .isr_vector section. The PICO doesn't have an sdcard, so there was alot more wasted space in there. By putting some object files which are always used in there, the PICO now fits in the 384K with about 2.7K of free space in .text

While fixing this, I also noticed that the STM32F401DISC is also broken. That board only has 256K of flash, and would require some more drastic changes like disabling the internal file system altogether. Currently storage.c (even before this patch) storage.c fails for that board.

If we want to fix that board, it should probably be done in a separate PR.

### CODE REVIEW

File: stmhal/boards/common.ld [hunk 1]
```diff
@@ -15,7 +15,11 @@ SECTIONS
 
         . = ALIGN(4);
         */ff.o(.text*)
-        */stm32f4xx_hal_sd.o(.text*)
+	*/vfs_fat_*.o(.text*)
+
+	*/py/formatfloat.o(.text*)
+	*/py/parsenum.o(.text*)
+	*/py/mpprint.o(.text*)
```
Comment (dpgeorge) [inline stmhal/boards/common.ld]: In the makefile there are rules to always use -Os for these files, which probably needs updating.

**Comment by dpgeorge** [comment]
> While fixing this, I also noticed that the STM32F401DISC is also broken.

Yes, that should be done separately to this PR.  But it's low priority :)

**Comment by profra** [comment]
@dhylands It means that my installation of uPY to Pico from last week is broken? ... and the build from today will be already OK?

**Comment by dhylands** [comment]
@profra It won't be fixed until this patch (or something similar) actually lands.

There have been a few changes recently which have had the size moving up and down and whether your build was broken would depend on the exact day it was grabbed.

**Comment by profra** [comment]
@dhylands It was this....
MicroPython v1.8.4-87-g53bfcc9 on 2016-09-28; Espruino Pico with STM32F401CD
Type "help()" for more information.
>>> 
Do you know if it is OK or not?

**Comment by dhylands** [comment]
@dpgeorge I added a comment to common.ld mentioning the -Os in the stmhal Makefile and updated the Makefile.

**Comment by dhylands** [comment]
@profra It really depends on the compiler used. I was able to download: http://micropython.org/resources/firmware/espruino-pico-20160909-v1.8.4.dfu and flashed it using pydfu:
```
775 >~/micropython/micropython/tools/pydfu.py -u espruino-pico-20160909-v1.8.4.dfu 
File: espruino-pico-20160909-v1.8.4.dfu
    DfuSe v1, image size: 273937, targets: 1
    Target 0, alt setting: 0, name: "ST...", size: 273652, elements: 2
      0, address: 0x08000000, size: 10020
      1, address: 0x08020000, size: 263616
    usb: 0483:df11, device: 0x0000, dfu: 0x011a, UFD, 16, 0x720b9219
Writing memory...
0x08000000   10020 [=========================] 100% 
0x08020000  263616 [=========================] 100% 
Exiting DFU...
```
What's significant is the 263616 number. Since there is only 384K flash that means that starting at 0x08020000 there is only 256K (or 262144 bytes) of flash left. 263616 is bigger than this, so there is 1472 bytes out of the image that weren't stored. Mine still booted to the REPL, but without a firmware map to know exactly what the missing data is, its hard to tell what's broken.

So, for your image, if that number is 262144 or less then your image is good. Otherwise something is broken.

**Comment by dhylands** [comment]
Another way to check: run the pyb.info() command and look for the _etext value. For the 1.8.4 image I get:
```
>>> pyb.info()
ID=38002900:13513430:36363435
S=84000000
H=84000000
P1=21000000
P2=42000000
_etext=80604b4
_sidata=80604b4
_sdata=20000000
_edata=2000010c
_sbss=2000010c
_ebss=20006544
_estack=20018000
_ram_start=20000000
_heap_start=20006544
_heap_end=20014000
_ram_end=20018000
qstr:
  n_pool=0
  n_qstr=0
  n_str_data_bytes=0
  n_total_bytes=0
GC:
  54656 total
  1136 : 53520
  1=21 2=6 m=16
LFS free: 2560 bytes
```
Since my _etext is 80604b4 which is bigger than 8060000 there is some portion of flash which wasn't programmed.

Now having said all that, I was able to read back my flash image and there was more than 384K stored there. Chip vendors will often do something called binning, where they'll take say a 512K flash part where some flash test failed in the 384K-512K range and call it a 384K part. Similar things happen with voltage, temperature, and frequency tests. Since they labelled it as a 'D' part it's only guaranteed that 384K works.

**Comment by profra** [comment]
@dhylands Thank you very much for your detailed explanations, I am again a bit smarter.
Here is my output... it means that I am also above the limit 806000. I have found the original bin file in my downloads folder and I would like ask you if my idea is right.... If I do verification i.e. compare this bin file with flash in Pico and the result is True then we can declare that uPy is fully functional?
```
 pyb.info()
ID=27006700:13513430:36363435
S=84000000
H=84000000
P1=21000000
P2=42000000
_etext=80600fc
_sidata=80600fc
_sdata=20000000
_edata=2000010c
_sbss=2000010c
_ebss=20006544
_estack=20018000
_ram_start=20000000
_heap_start=20006544
_heap_end=20014000
_ram_end=20018000
qstr:
  n_pool=0
  n_qstr=0
  n_str_data_bytes=0
  n_total_bytes=0
GC:
  54656 total
  368 : 54288
  1=12 2=4 m=3
LFS free: 41984 bytes
```

**Comment by dhylands** [comment]
@profra 

> If I do verification i.e. compare this bin file with flash in Pico and the result is True then we can declare that uPy is fully functional?

Well it's fully functional on your Espruino Pico board, but not necessarily everybody else's Espruino Pico board. Using anything above 384K is basically a crap shoot (which is why this PR changes things to make the linker fail if the code gets too big). My pico board works with 1.8.4 but some newer images I tried fail.

Once this PR gets merged, I'll look into turning off the software I2C & SPI support (since I saw those being linked in my map file) and that should get some more breathing room for core changes.

**Comment by dpgeorge** [comment]
Thanks Dave, merged in 7a9c183c20f01eec204b7bb89180aebe9723e576 with the following fixes: xe flash text size was still 512k (changed to 384k); xd flash total size was still 512k, changed to 384k; added flash_fs sections

**Pull request closed without merging**